### PR TITLE
xfd, tag: add "desired outcome" dropdown (with delete, merge, redirect, draftify), and remove {{Merge}} tags from tag module

### LINKED ISSUE #2335: xfd: add support for preferred outcome (delete, merge, redirect, draftify) at AFD
Requested by Voorts at https://en.wikipedia.org/wiki/Wikipedia_talk:Twinkle#Updates_to_AfD_backend

Mostly copy/pasted from https://github.com/wikimedia-gadgets/twinkle/issues/2335#issuecomment-4188615635, with thanks to FaviFake:

* [x] Update XFD -> AFD to add a dropdown that lets the user select their preferred outcome, and also do some additional things based on what is selected:
  * **Delete**
  * **Merge**
    1. On the nominated page:
       * `{{Article for deletion/dated |outcome=merging |...}}`
  * **Redirect**
    1. On the nominated page:
       * `{{Article for deletion/dated |outcome=redirecting |...}}`
  * **Draftify**
    1. On the nominated page:
       * `{{Article for deletion/dated |outcome=draftification |...}}`
* [x] **Draftify**
    1. On the author's user talk page, place this:
       * `{{subst:Afd notice |1=SourcePage |outcome=draftification}}`
    2. Pre-fill some draftify text in the AFD reason textarea that is published on the AFD page, i.e. "Propose '''draftifying'''
* [x] **Redirect**
    1. Collect "TargetPage" in a text box
    2. On the nominated page:
       * `{{Article for deletion/dated |target=TargetPage |...}}`
    3. On the author's user talk page:
       * if "TargetPage" was given, place this:
         * `{{subst:Afd notice |1=SourcePage |target=TargetPage |outcome=redirecting}}`
       * if "TargetPage" was not given, place this:
         * `{{subst:Afd notice |1=SourcePage |outcome=redirecting}}`
    4. Pre-fill some redirect text in the AFD reason textarea that is published on the AFD page, i.e. "Propose '''redirecting''' [[]] to [[]]"
* [x] **Merge**
    1. Collect "TargetPage" in a text box
    2. On the nominated page:
       * `{{Article for deletion/dated |target=TargetPage |...}}`
    3. if "TargetPage" was given, place this on the target page (always include NominationName):
        * `{{Merge from |1=SourcePage |target=TargetPage |afd=NominationName |date ={{subst:CURRENTMONTHNAME}} {{subst:CURRENTYEAR}} }}`
    4. On the author's user talk page:
       * if "Target" was given, place this:
         * `{{subst:Afd notice |1=SourcePage |target=TargetPage |outcome=merging}}`
       * if "Target" was not given, place this:
         * `{{subst:Afd notice |1=SourcePage |outcome=merging}}`
    5. Pre-fill some merge text in the AFD reason textarea that is published on the AFD page, i.e. "Propose '''merging''' [[]] to [[]]"
* [ ] in the tag module, delete...
  * {{Merge}}
  * {{Merge to}}
  * {{Merge from}}
* [ ] one final manual test of everything, including of the edge cases of the target page being protected or not existing

Up next: update XFDcloser: https://github.com/wikimedia-gadgets/xfdcloser/issues/135

### PR DESCRIPTION
**Pull request by NovemLinguae**
Why
- a recent RFC decided to bring the merging process into the AFD process

What
- xfd module
    - add dropdown called "Desired outcome:" with options Delete, Merge, Redirect, and Draftify
    - sometimes add text box "Target page:" depending on what is chosen in the dropdown
    - for merging, when a target page is provided, add a {{Merge from}} tag to the target page
    - modifications to outputted wikicode (of the article tag and the user talk notice)
    - add tests
    - minor refactoring of some unrelated code
- tag module
    - remove {{Merge}}, {{Merge to}}, and {{Merge from}}, since gnomes are currently manually converting all these to AFDs. will save them some work. folks need to use the new process.
- a more detailed list of changes can be found in the ticket: #2335

Fixes #2335

---

New dropdown:

New target page input box, new default reason:

Warning message when target page doesn't exist (non-fatal error):
image

### CODE REVIEW

File: src/modules/twinklexfd.js [hunk 1]
```diff
@@ -750,6 +763,36 @@ Twinkle.xfd.callback.change_category = function twinklexfdCallbackChangeCategory
...
 Twinkle.xfd.callbacks = {
+	changeOutcome: function(outcome) {
+		const form = outcome.target.form;
+		const reasonBox = form.querySelector('textarea[name="reason"]');
+		const targetInput = form.target;
+		if (outcome.target.value === 'redirecting') {
+			reasonBox.value = "I propose '''redirecting''' because ";
+			// add target text box
+			if (!targetInput) {
+				const targetElem = new Morebits.QuickForm.Element({
```
Comment (siddharthvp) [inline src/modules/twinklexfd.js]: instead of physically adding/removing the input, we can create it the normal way and just control its display/disabled attributes.

File: src/modules/twinklexfd.js [hunk 2]
```diff
@@ -750,6 +763,36 @@ Twinkle.xfd.callback.change_category = function twinklexfdCallbackChangeCategory
 };
 
 Twinkle.xfd.callbacks = {
+	changeOutcome: function(outcome) {
+		const form = outcome.target.form;
+		const reasonBox = form.querySelector('textarea[name="reason"]');
```
Comment (siddharthvp) [inline src/modules/twinklexfd.js]: can also be written as `form.reason`

File: src/modules/twinklexfd.js (hunk 1 above)
Comment (NovemLinguae) [reply src/modules/twinklexfd.js]: I tried a couple different things and couldn't get it to work. Here's some things I tried:

```
			target = work_area.append({
				name: 'target',
				type: 'input',
				label: 'Target page:',
				tooltip: 'Target page for the redirect.',
				// Hide this initially. Unhide it later for certain outcomes, using a callback.
				style: 'visibility: hidden;'
			});
```

```
			target = work_area.append({
				name: 'target',
				type: 'input',
				label: 'Target page:',
				tooltip: 'Target page for the redirect.'
			});
			// Hide this initially. Unhide it later for certain outcomes, using a callback.
			Morebits.QuickForm.setElementVisibility( $(target).parent(), false);
```

One problem I had is that append would add a div node containing a label and an input, but styling is only applied to the input. Also `Morebits.QuickForm.setElementVisibility(` wasn't working... maybe I was doing something wrong.

Any suggestions?

Comment (siddharthvp) [reply src/modules/twinklexfd.js]: The second snippet doesn't work since `target` is a quickform element, not an actual DOM node. It also doesn't yet exist on the DOM. It needs to be made hidden after the form has been rendered.

**Comment by NovemLinguae** [comment]
I merged all patches together into this patch. This is ready for final review.

File: src/modules/twinkletag.js [hunk 3]
```diff
@@ -1693,15 +1588,12 @@ Twinkle.tag.callbacks = {
 					tags.push(tag);
 				}
 			} else {
-				if (tag === 'Merge from' || tag === 'History merge') {
+				if (tag === 'History merge') {
 					tags.push(tag);
 				} else {
 					Morebits.Status.warn('Info', 'Found {{' + tag +
 						'}} on the article already...excluding');
 					// don't do anything else with merge tags
```
Comment (siddharthvp) [inline src/modules/twinkletag.js]: comment can be removed

File: src/modules/twinklexfd.js [hunk 4]
```diff
@@ -1179,11 +1236,20 @@ Twinkle.xfd.callbacks = {
...
 
-			params.tagText = (params.noinclude ? '<noinclude>{{' : '{{') + (params.number === '' ? 'subst:afd|help=off' : 'subst:afdx|' +
-					params.number + '|help=off') + (params.noinclude ? '}}</noinclude>\n' : '}}\n');
+			// Add AFD tag to article
+			params.tagText = Twinkle.xfd.callbacks.afd.generateArticleTagWikitext(
+				params.noinclude, params.outcome, params.afdtarget, params.number
+			);
+
+			// If the selected outcome is merge, add {{Merge from}} to the target page
+			if ( params.outcome === 'merging' && params.afdtarget ) {
```
Comment (siddharthvp) [inline src/modules/twinklexfd.js]: extra spaces in parens

File: src/modules/twinklexfd.js [hunk 5]
```diff
@@ -919,25 +962,7 @@ Twinkle.xfd.callbacks = {
...
-				}
-				break;
-			case 'cfd':
-				notifytext += '|action=' + params.action + (mw.config.get('wgNamespaceNumber') === 10 ? '|stub=yes' : '');
-				break;
-			default: // ffd, rfd
-				break;
-		}
-		notifytext += '|1=' + Morebits.pageNameNorm + '}} ~~~~';
+		const notifyText = Twinkle.xfd.callbacks.generateUserTalkNoticeWikitext(params.venue, params.outcome, params.afdtarget, params.numbering, params.xfdcat, params.tfdtarget, params.action, mw.config.get('wgNamespaceNumber'), Morebits.pageNameNorm);
```
Comment (siddharthvp) [inline src/modules/twinklexfd.js]: We can pass the `params` object directly. Passing the namespace number and page name looks awkward as they are globals. It's probably better to mock them in the tests.

File: tests/twinklexfd.test.js [hunk 6]
```diff
@@ -109,4 +109,181 @@ describe('modules/twinklexfd', () => {
...
+		test('afd, deletion, 1st nomination', () => {
+			const params = {};
+			params.venue = 'afd';
+			params.outcome = 'deletion';
+			params.afdtarget = '';
+			params.numbering = '';
+			params.xfdcat = '?';
+			params.tfdtarget = undefined;
+			params.action = undefined;
+			mw.config.get = (key) => key === 'wgNamespaceNumber' ? 0 : undefined;
```
Comment (siddharthvp) [inline tests/twinklexfd.test.js:198]: can be written as `mw.config.set('wgNamespaceNumber', 0)`

File: src/modules/twinklexfd.js [hunk 7]
```diff
@@ -750,7 +773,26 @@ Twinkle.xfd.callback.change_category = function twinklexfdCallbackChangeCategory
...
 Twinkle.xfd.callbacks = {
-	// Requires having the tag text (params.tagText) set ahead of time
+	changeAfdOutcome: function(outcome) {
+		const $reason = $('[name="reason"]');
+		const $afdTarget = $('[name="afdtarget"]');
+		if (outcome.target.value === 'redirecting') {
+			$reason.val("I propose '''redirecting''' to [[YOUR TARGET ARTICLE HERE]] because ");
+			$afdTarget.parent().show();
+		} else if (outcome.target.value === 'merging') {
+			$reason.val("I propose '''merging''' to [[YOUR TARGET ARTICLE HERE]] because ");
```
Comment (siddharthvp) [inline src/modules/twinklexfd.js:783]: If they submit without replacing it, can we substitute it with params.afdtarget from our side? I suspect some folks may submit as-is expecting it to be auto-substituted.

Comment (NovemLinguae) [reply src/modules/twinklexfd.js]: Done. I decided to do this when they type in the afdtarget box rather than when they submit the form. There is a small risk we miss it if they type in the reason box before they type in the target box, but I think this is acceptable.

**Comment by NovemLinguae** [comment]
I plan to self merge this on Monday unless there's objections. Twinkle is currently placing non-ideal wikicode that needs manual correction by gnomes, and this patch would fix it, reducing their workload.

**Comment by NovemLinguae** [comment]
I'm not sure why Copilot thinks it's a co-author of this huge patch I spent a lot of time writing. That's pretty annoying. GitHub's marketing department and robots shouldn't be permitted to insert stuff like this without even prompting me.

image

**Comment by siddharthvp** [comment]
Seems to have come from the last commit which had `Co-authored-by: Copilot <copilot@github.com>` and was kept while squashing. I always clean up the commit message while squashing instead of keeping the full history of intermediate commits.

**Comment by NovemLinguae** [comment]
After some digging I figured out it is a recent change to the VS Code GitHub Copilot plugin, and can be disabled in VS Code settings.
